tests/client/coll.rs

```diff
@@ -59,22 +53,308 @@ fn find_and_insert_one() {
+
+    coll.insert_many(vec![doc1.clone(), doc2.clone(), doc3.clone()], false, None)
+        .ok().expect("Failed to insert documents into collection.");
+
+    // Update single document
+    let update = doc! {
+        "$set" => nested_doc! {
+            "director" => Bson::String("Robert Zemeckis".to_owned())
+        }
     };
 
+    let result = coll.find_one_and_update(doc2.clone(), update, None)
+        .ok().expect("Failed to execute find_one_and_update command.");
+
+    match result.unwrap().get("title") {
+        Some(&Bson::String(ref title)) => assert_eq!("Back to the Future", title),
+        _ => panic!("Expected Bson::String!"),
+    }
+
+    // Validate state of collection
+    let mut cursor = coll.find(None, None).ok().expect("Failed to execute find command.");
+    let results = cursor.next_n(3);
+    assert_eq!(3, results.len());
+
+    // Assert director attributes
+    assert!(results[0].get("director").is_none());
+    assert!(results[2].get("director").is_none());
+    match results[1].get("director") {
+        Some(&Bson::String(ref director)) => assert_eq!("Robert Zemeckis", director),
+        _ => panic!("Expected Bson::String!"),
+    }
+}
+
+#[test]
+fn aggregate() {
+    let client = MongoClient::with_uri("mongodb://localhost:27017").unwrap();
+    let db = client.db("test");
+    let coll = db.collection("aggregate");
+
+    db.drop_database().ok().expect("Failed to drop database");
+
+    // Insert documents
+    let tags1 = vec![Bson::String("a".to_owned()),
+                     Bson::String("b".to_owned()),
+                     Bson::String("c".to_owned())];
+
+    let tags2 = vec![Bson::String("a".to_owned()),
+                     Bson::String("b".to_owned()),
+                     Bson::String("d".to_owned())];
+
+    let tags3 = vec![Bson::String("d".to_owned()),
+                     Bson::String("e".to_owned()),
+                     Bson::String("f".to_owned())];
+
+    let doc1 = doc! { "tags" => Bson::Array(tags1) };
+    let doc2 = doc! { "tags" => Bson::Array(tags2) };
+    let doc3 = doc! { "tags" => Bson::Array(tags3) };
+
+    coll.insert_many(vec![doc1.clone(), doc2.clone(), doc3.clone()], false, None)
+        .ok().expect("Failed to execute insert_many command.");
+
+    // Build aggregation pipeline to count instances of each letter in tag arrays
+    let project = doc! { "$project" => nested_doc! { "tags" => Bson::I32(1) } };
+    let unwind = doc! { "$unwind" => Bson::String("$tags".to_owned()) };
+    let group = doc! { "$group" => nested_doc! {
+        "_id" => Bson::String("$tags".to_owned()),
+        "count" => nested_doc! { "$sum" => Bson::I32(1) }
+    } };
+
+    // Aggregate
+    let mut cursor = coll.aggregate(vec![project, unwind, group], None)
+        .ok().expect("Failed to execute aggregate command.");
+
+    let results = cursor.next_n(10);
+    assert_eq!(6, results.len());
+
+    for i in 0..6 {
```

Comment:
Let's not do complex for loops in tests, that just makes test failures very difficult to debug